xc/device.ml

```diff
@@ -1143,18 +1143,67 @@ module PCI = struct
     (* Sort into the order the devices were plugged *)
     List.sort (fun a b -> compare (fst a) (fst b)) pairs
 
-  let add ~xs pcidevs domid =
+  let _pci_add ~xc ~xs pcidev domid slot =
+    let sysfs_pci_dev = "/sys/bus/pci/devices/" in
+    let bus = 0 in
+    let func = 0 in
+    let pci_bdf = Printf.sprintf "%04x:%02x:%02x.%01x" domid bus slot func in
+    let page_size = 4096 in
+    if (Qemu.is_running ~xs domid) then
+      begin
+        let _qmp_result = qmp_send_cmd domid
+            (Qmp.Device_add {driver="xen-pci-passthrough";
+                             device=Qmp.Device.PCI({id=pcidev; dev=0; fn=0; hostaddr=pcidev; permissive=false})}) in
+        Unixext.with_input_channel (sysfs_pci_dev ^ pci_bdf ^ "/resource") (fun in_channel ->
+            for i = 0 to 7 do
+              let start = ref 0 in
+              let end_ = ref 0 in
+              let flags = ref 0 in
+              let size = ref 0 in
+              Scanf.bscanf (Scanf.Scanning.from_channel in_channel) "0x%llx 0x%llx 0x%llx\n" (fun scan_start scan_end scan_flags ->
+                  start := scan_start;
+                  end_ := scan_end;
+                  flags := scan_flags);
+              size := !end_ - !start + 1;
+              if !start <> 0 then
+                if !flags <> 0 then
+                  Xenctrl.domain_iomem_permission xc domid (Nativeint.of_int !start) (Nativeint.of_int !size) true
+                else
+                  Xenctrl.domain_iomem_permission xc domid (Nativeint.of_int (!start lsr 12)) (Nativeint.of_int ((page_size + !size -1) lsr 12)) true
+            done
+          );
+        Unixext.with_input_channel (sysfs_pci_dev ^ pci_bdf) (fun in_channel ->
+            let irq = Scanf.bscanf (Scanf.Scanning.from_channel in_channel) "%u" (fun i -> i) in
+            let irq = Xenctrlext.physdev_map_pirq xc domid (Int64.of_int irq) in
+            Xenctrl.domain_irq_permission xc domid irq true
+          )
+      end
+
+  let add ~xc ~xs pcidevs domid =
     try
-      let current = read_pcidir ~xs domid in
-      let next_idx = List.fold_left max (-1) (List.map fst current) + 1 in
-      add_xl pcidevs domid;
-      List.iteri
-        (fun count address ->
-           xs.Xs.write
-             (device_model_pci_device_path xs 0 domid ^ "/dev-" ^ string_of_int (next_idx + count))
-             (Xenops_interface.Pci.string_of_address address))
-        pcidevs
-    with exn -> raise (Cannot_add (pcidevs, exn))
+      if !Xenopsd.use_old_pci_add then
+        begin
+          let current = read_pcidir ~xs domid in
+          let next_idx = List.fold_left max (-1) (List.map fst current) + 1 in
+          add_xl (List.rev_map (fun (a,_) -> a) pcidevs) domid;
+          List.iteri
+            (fun count (address, _) ->
+               xs.Xs.write
+                 (device_model_pci_device_path xs 0 domid ^ "/dev-" ^ string_of_int (next_idx + count))
```

Comment:
Use `Printf.sprintf`

Reply:
I will do, but in looking at this function call closer I can't work it out, it seems to be missing a parameter from the middle? It's meant to be `xs -> domid -> int -> string`. But with `^` after the domid, that whole last part is the string arg. So where is the int or domid param?

Reply:
Note that the type `domid` is also an `int`. The `domid` and `int` args of `device_model_pci_device_path` are in fact both domids: first for the backend domain, next for the frontend.